src/vs/workbench/common/editor/editorLabels.ts

```diff
@@ -0,0 +1,188 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Emitter, Event } from 'vs/base/common/event';
+import { match } from 'vs/base/common/glob';
+import { Disposable } from 'vs/base/common/lifecycle';
+import { basename, dirname, extname, isAbsolutePath, relativePath } from 'vs/base/common/resources';
+import { URI } from 'vs/base/common/uri';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
+import { IWorkspaceContextService, IWorkspaceFolder } from 'vs/platform/workspace/common/workspace';
+
+interface IFilenameAttributes {
+	// grand-parent/parent/filename.ext1.ext2 -> parent
+	dirname: string;
```

Comment:
I tend to set properties `readonly` where possible, not sure thats your style too.